### PR TITLE
feat: Tensorboard RBAC

### PR DESCRIPTION
**Pull request by eecsliu**
## Description
Basic tensorboard implementation of RBAC for OSS.

## Commentary (optional)

DET-8781

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[DET-8781]: https://determinedai.atlassian.net/browse/DET-8781?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hamidzr — COMMENTED** [review]
looks good. some quick comments. It'll take me a bit longer to look more closely. do we want to also call into this new interface in some other places we currently use NSTAuthZ because we didn't have one for TSB? eg command/events and api_auth

File: master/internal/command/authz_tensorboard_iface.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/determined-ai/determined/master/internal/authz"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// TensorboardAuthZ describes authz methods for Tensorboards.
+type TensorboardAuthZ interface {
+	// GET /api/v1/tensorboards/:tb_id
+	CanGetTensorboard(
+		ctx context.Context, curUser *model.User, tb *tensorboardv1.Tensorboard,
+	) (canGetTensorboard bool, serverError error)
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/command/authz_tensorboard_iface.go:17]: Right sorry I forgot to follow up on that. I checked and I didn't see any real reason to keep it, I was going to check with rbac team.

File: master/internal/command/tensorboard_manager.go [hunk 2]
```diff
@@ -43,7 +43,8 @@ func (t *tensorboardManager) Receive(ctx *actor.Context) error {
 		}
 		for _, tensorboard := range ctx.AskAll(&tensorboardv1.Tensorboard{}, ctx.Children()...).GetAll() {
 			typed := tensorboard.(*tensorboardv1.Tensorboard)
-			if (len(users) == 0 && len(userIds) == 0) || users[typed.Username] || userIds[typed.UserId] {
+			if ((len(users) == 0 && len(userIds) == 0) || users[typed.Username] || userIds[typed.UserId]) &&
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/command/tensorboard_manager.go:46]: ok will change

File: master/internal/api_tensorboard.go [hunk 3]
```diff
@@ -124,8 +117,20 @@ func (a *apiServer) GetTensorboard(
 func (a *apiServer) KillTensorboard(
 	ctx context.Context, req *apiv1.KillTensorboardRequest,
 ) (resp *apiv1.KillTensorboardResponse, err error) {
-	if _, err := a.GetTensorboard(ctx,
-		&apiv1.GetTensorboardRequest{TensorboardId: req.TensorboardId}); err != nil {
+	getResponse, err := a.GetTensorboard(ctx,
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/api_tensorboard.go:120]: I'm not sure I see what you mean - isn't the check that happens in `GetTensorboard` supposed to check for getting experiment artifacts?

File: master/internal/command/tensorboard_manager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/command/tensorboard_manager.go:46]: should have qualified it with "very nit-picky" works either way:)

File: master/internal/api_tensorboard.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api_tensorboard.go:120]: true I was thinking about it in terms of other NSC.
now that we do have separate RBAC interfaces for these different actions maybe the check for "can you terminate a tensorboard" would fail at `CanTerminateTensorboard` instead of `canGetTensorboard` and then the internal implementation could refer to `CanGetTensorboard` this would probably lead to better auditable logs 🤔

File: master/internal/command/authz_tensorboard_iface.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+	// GET /api/v1/tensorboards
+	FilterTensorboards(
```
Comment (hamidzr) [inline master/internal/command/authz_tensorboard_iface.go:20]: we only really need the workspace ids of the tensorboards here. what you have here feels natural but how do we keep it consistent with the NCSAuthZ iface if we wanted to? 
 https://github.com/determined-ai/determined/pull/5609/files#diff-5306290d2cf5fbcbbf54ff35798d9a111e710c775399f4f09c996079fb8f98c8R27

**Review by hamidzr — APPROVED** [review]

File: master/internal/api_tensorboard.go [hunk 5]
```diff
@@ -112,8 +105,8 @@ func (a *apiServer) GetTensorboard(
 		return nil, err
 	}
 
-	if ok, err := command.AuthZProvider.Get().CanGetNSC(
-		ctx, *curUser, model.UserID(resp.Tensorboard.UserId), command.PlaceHolderWorkspace); err != nil {
+	if ok, err := command.TbAuthZProvider.Get().CanGetTensorboard(
```
Comment (ioga) [inline master/internal/api_tensorboard.go:108]: Original intention was to have a single provider per module, i.e. one `command.AuthZProvider` for all "command" stuff.

Comment (hamidzr) [reply master/internal/api_tensorboard.go:108]: Yeah, and for the other NSCs we have a single interface but since the tensorboards case is a bit different, where it needs to consult with the experiment authz interface, this is introduced as one of the options. We (me and @eecsliu ) chatted about an alternative where I think we can delegate all of the tensorboard authorizations to `CanGetExperimentArtifacts` and maybe introduce a new method to do this in batch where we can use it for all the sources of a tensorbaord in one go. Eric would probably have more to add.

Comment (ioga) [reply master/internal/api_tensorboard.go:108]: but the methods on this interface can just be a part of methods of the shared "command" interface. it does not matter really if they have a different signature. using stuff from experiments could've been a deciding factor if it cause an import cycle, but then you'd have to split this go module anyway.

Comment (eecsliu) [reply master/internal/api_tensorboard.go:108]: ah ok I'll just merge this into the commands interface

File: master/internal/api_tasks_intg_test.go [hunk 6]
```diff
@@ -25,9 +25,7 @@ func errTaskNotFound(id string) error {
 	return status.Errorf(codes.NotFound, "task not found: %s", id)
 }
 
-var (
-	authzCommand *mocks.NSCAuthZ
-)
+var authzCommand *mocks.NSCAuthZ
```
Comment (dannysauer) [inline master/internal/api_tasks_intg_test.go:28]: Correct me if this isn't just cosmetic, but it kinda feels like this sort of cleanup oughtta be in a separate PR.  Same with the `var id =` to `id :=` changes. :)

Comment (hamidzr) [reply master/internal/api_tasks_intg_test.go:28]: agreed. this should be a noOp here anyway the target branch [already has it](https://github.com/determined-ai/determined/commit/7d9f25538e40ebb0ec193298f11656a4da0c8aaf) not sure why githug ui wouldn't show an updated view of it

File: master/internal/command/authz_tensorboard_iface.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/command/authz_tensorboard_iface.go:17]: consulted @maxrussell about this. TLDR is that our underlying implementation just uses a special error so we _could_ simplify the interface, but from an api design standpoint it's a good idea to be clear what is a permission granted/denied and what is a server error. From a golang standpoint, this makes sense and is pretty typical, such as having something like `ok, err := fn()`.

File: master/internal/api_tasks_intg_test.go (hunk 6 above)
Comment (eecsliu) [reply master/internal/api_tasks_intg_test.go]: rebased and this is an issue no longer!

File: master/internal/command/authz_iface.go [hunk 7]
```diff
@@ -32,6 +34,27 @@ type NSCAuthZ interface {
...
+
+	// POST /api/v1/tensorboards/:tb_id/kill
+	CanTerminateTensorboard(
+		ctx context.Context, curUser *model.User, tb *tensorboardv1.Tensorboard,
+	) error
+
+	// POST /api/v1/tensorboards/:tb_id/set_priority
+	CanSetTensorboardPriority(
+		ctx context.Context, curUser *model.User, tb *tensorboardv1.Tensorboard,
+	) error
```
Comment (hamidzr) [inline master/internal/command/authz_iface.go:57]: do we want these to take in the whole tensorboardv1.Tensorboard or stay closer to what the NSC interface is asking for? aka workspaceId. Imo these could be more like the existing NSC ones

File: master/internal/command/authz_basic_impl.go [hunk 8]
```diff
@@ -42,6 +45,40 @@ func (a *NSCAuthZBasic) CanSetNSCsPriority(
 	return nil
 }
 
+// CanGetTensorboard returns true and nil error unless the developer master config option
+// security.authz._strict_ntsc_enabled is true then it returns a boolean if the user is
+// an admin or if the user owns the tensorboard and a nil error.
+func (a *NSCAuthZBasic) CanGetTensorboard(
+	ctx context.Context, curUser *model.User, tb *tensorboardv1.Tensorboard,
+) (canGetTensorboards bool, serverError error) {
+	if !config.GetMasterConfig().Security.AuthZ.StrictNTSCEnabled {
```
Comment (hamidzr) [inline master/internal/command/authz_basic_impl.go:54]: since https://github.com/determined-ai/determined/pull/5639  has landed I don't think we want this logic here anymore. just return nil?

File: proto/src/determined/api/v1/tensorboard.proto [hunk 9]
```diff
@@ -40,6 +40,9 @@ message GetTensorboardsRequest {
   // Limit tensorboards to those that are owned by users with the specified
   // userIds.
   repeated int32 user_ids = 6;
+  // Limit tensorboards to those that are in a specific workspace, or 0 for
```
Comment (hamidzr) [inline proto/src/determined/api/v1/tensorboard.proto:43]: add the sort parameter as well?  (to sort by workspace id)

File: master/internal/api_tensorboard.go [hunk 10]
```diff
@@ -112,8 +105,8 @@ func (a *apiServer) GetTensorboard(
 		return nil, err
 	}
 
-	if ok, err := command.AuthZProvider.Get().CanGetNSC(
-		ctx, *curUser, model.UserID(resp.Tensorboard.UserId), command.PlaceHolderWorkspace); err != nil {
+	if ok, err := command.AuthZProvider.Get().CanGetTensorboard(
```
Comment (hamidzr) [inline master/internal/api_tensorboard.go:108]: should we add tensorboards to these tests as well https://github.com/hamidzr/determined/blob/2e16fc41f2a98a7afb8a124f1ad92b8d43c169e5/master/internal/api_ntsc_intg_test.go#L78 ? we could add them in a separate pr

**Review by hamidzr — CHANGES_REQUESTED** [review]
thanks for updating this. two questions
- do we know what the scenario would look like when we potentially start pulling in from the experiment package into the command package to check if use has access to tsb source experiments? maybe we can somehow avoid this?
- wdy think about keeping the interface input type closer to the NSC ones? (I added an inline comment)

**Comment by hamidzr** [comment]
probably want to account for https://determined-ai.slack.com/archives/C03LE5Z8R46/p1672776613155459?thread_ts=1672772755.830529&cid=C03LE5Z8R46 as well. maybe something like https://github.com/determined-ai/determined/pull/5609/commits/7b659491a2f804bc267f9f32e6b3412846571b82#diff-5306290d2cf5fbcbbf54ff35798d9a111e710c775399f4f09c996079fb8f98c8R27-R30 could work

File: master/internal/command/authz_iface.go (hunk 7 above)
Comment (eecsliu) [reply master/internal/command/authz_iface.go:57]: my bad, I misread the code - I thought that it would require another database query, so I decided to keep it this way in order to minimize the queries. I just looked more closely and it appears workspace ID works. Will spend a few more minutes to see what I can combine with the NSC interfaces.

File: master/internal/command/authz_basic_impl.go (hunk 8 above)
Comment (eecsliu) [reply master/internal/command/authz_basic_impl.go:54]: Ah right, good catch

File: master/internal/api_tensorboard.go (hunk 10 above)
Comment (eecsliu) [reply master/internal/api_tensorboard.go:108]: I was going to add them to a separate PR - I also was going to rename NSC to NTSC to include tensorboards, so I can do a follow up PR so this one doesn't get too large

File: master/internal/command/authz_tensorboard_iface.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/command/authz_tensorboard_iface.go]: > It's a good idea to be clear what is a permission granted/denied and what is a server error.

I think we definitely want regardless of how we get there.

> From a golang standpoint, this makes sense 

by "this" do you mean `bool, err` return type instead of `permErr, serverErr error` ? or the other way

**Comment by eecsliu** [comment]
> we probably also want to account for https://determined-ai.slack.com/archives/C03LE5Z8R46/p1672776613155459?thread_ts=1672772755.830529&cid=C03LE5Z8R46 as well. maybe something like [7b65949#diff-5306290d2cf5fbcbbf54ff35798d9a111e710c775399f4f09c996079fb8f98c8R27-R30](https://github.com/determined-ai/determined/commit/7b659491a2f804bc267f9f32e6b3412846571b82#diff-5306290d2cf5fbcbbf54ff35798d9a111e710c775399f4f09c996079fb8f98c8R27-R30) could work

For the basic RBAC I don't think we need to do any filtering, but I ended up following a path similar to what you did in `7b65949`

**Comment by hamidzr** [comment]
> > we probably also want to account for https://determined-ai.slack.com/archives/C03LE5Z8R46/p1672776613155459?thread_ts=1672772755.830529&cid=C03LE5Z8R46 as well. maybe something like [7b65949#diff-5306290d2cf5fbcbbf54ff35798d9a111e710c775399f4f09c996079fb8f98c8R27-R30](https://github.com/determined-ai/determined/commit/7b659491a2f804bc267f9f32e6b3412846571b82#diff-5306290d2cf5fbcbbf54ff35798d9a111e710c775399f4f09c996079fb8f98c8R27-R30) could work
> 
> For the basic RBAC I don't think we need to do any filtering, but I ended up following a path similar to what you did in `7b65949`

yeah we don't need to have it. the main motivations for me about this have been to minimize potential code change conflicts and try to come up with interfaces that allow us to keep the api_* files unchanged between ee and oss and encapsulate the differences to the interface implementation

File: master/internal/command/authz_basic_impl.go (hunk 8 above)
Comment (hamidzr) [reply master/internal/command/authz_basic_impl.go:59]: did you mean to remove this?

File: master/internal/api_tensorboard.go [hunk 11]
```diff
@@ -80,20 +105,8 @@ func (a *apiServer) GetTensorboards(
...
-			ctx, *curUser, model.UserID(resp.Tensorboards[i].UserId), command.PlaceHolderWorkspace)
-		if serverError != nil {
-			err = serverError
-		}
-		return ok
-	})
-	if err != nil {
-		return nil, err
-	}
+	filteredTensorboards, err := a.filterTensorboards(ctx, *curUser, resp.Tensorboards)
```
Comment (hamidzr) [inline master/internal/api_tensorboard.go:108]: It seems like there is a missing piece on looking at what the user requested  `req.workspaceId` (something we might want to change into a list of workspace ids at some point)

or is this what you intended to do at a later pr? I think this affects how `AccessibleScopes` is defined so IMO it makes sense to define Accessible scopes at the same time that you start looking at this parameter. aka if you prefer to exclude `req.workspaceId` from this pr take out defining the `AccessibleScope` as well and add them together in a different pr

File: master/internal/command/authz_iface.go [hunk 12]
```diff
@@ -36,25 +36,19 @@ type NSCAuthZ interface {
...
 
-	// GET /api/v1/tensorboards
-	FilterTensorboards(
-		ctx context.Context, curUser *model.User, tensorboards []*tensorboardv1.Tensorboard,
-	) ([]*tensorboardv1.Tensorboard, error)
-
 	// POST /api/v1/tensorboards/:tb_id/kill
 	CanTerminateTensorboard(
-		ctx context.Context, curUser *model.User, tb *tensorboardv1.Tensorboard,
+		ctx context.Context, curUser model.User, tb *tensorboardv1.Tensorboard,
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/command/authz_iface.go:51]: oh yeah nice catch...this should have also been changed but idk what happened

File: master/internal/api_tensorboard.go [hunk 13]
```diff
@@ -68,6 +68,31 @@ func filesToArchive(files []*utilv1.File) archive.Archive {
...
+
+	filteredScopes, err := command.AuthZProvider.Get().AccessibleScopes(
+		ctx, curUser, scopes)
+	if err != nil {
+		return nil, err
+	}
+
+	var filteredTensorboards []*tensorboardv1.Tensorboard
+
+	for _, tb := range tensorboards {
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/api_tensorboard.go:87]: I'll check it out

[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/api_tensorboard.go:87]: Tried compressing the logic into `a.filter` and it doesn't help much. I think the existing way is a little clearer as to what is happening. I think `a.filter` would be more useful if I was operating inside of the api handler and had not already moved this logic into its own function

File: master/internal/api_tensorboard.go [hunk 14]
```diff
@@ -80,20 +103,9 @@ func (a *apiServer) GetTensorboards(
...
-			ctx, *curUser, model.UserID(resp.Tensorboards[i].UserId), command.PlaceHolderWorkspace)
-		if serverError != nil {
-			err = serverError
-		}
-		return ok
-	})
-	if err != nil {
-		return nil, err
-	}
+	filteredTensorboards, err := a.filterTensorboards(ctx, *curUser, resp.Tensorboards,
```
Comment (hamidzr) [inline master/internal/api_tensorboard.go:106]: do we need to check this error? 
question: also would this return a 404 if the user requests workspace that exists but they don't have access to?

**Review by hamidzr — APPROVED** [review]
woohoo! 🚢

File: master/internal/command/authz_basic_impl.go [hunk 15]
```diff
@@ -42,6 +43,44 @@ func (a *NSCAuthZBasic) CanSetNSCsPriority(
...
 }
 
+// AccessibleScopes returns the set of scopes that the user should be limited to.
+func (a *NSCAuthZBasic) AccessibleScopes(
+	ctx context.Context, curUser model.User, requestedScope model.AccessScopeID,
+) (model.AccessScopeSet, error) {
+	var ids []int
+	returnScope := model.AccessScopeSet{}
+
+	if requestedScope == 0 {
```
Comment (hamidzr) [inline master/internal/command/authz_basic_impl.go:53]: I think there is a case to be made here that we shouldn't need to go to the db in this case in the basic impl since we know we're not limiting the user. but the other options has their own downsides as well. this works for now imo

Comment (eecsliu) [reply master/internal/command/authz_basic_impl.go:53]: yeah imo the best way to implement this function is have it return all the workspaces you have access to. That way there is no ambiguity between special values like nil or zero.

File: master/internal/api_tensorboard.go (hunk 14 above)
Comment (eecsliu) [reply master/internal/api_tensorboard.go:106]: doh good catch, was trying to be quick but was sloppy. The basic implementation will not restrict the user in any way, so it should never return any error. For EE, it depends on how we implement the function. I think we should just return an empty list if they don't have access to a workspace.

Comment (hamidzr) [reply master/internal/api_tensorboard.go:106]: related conversation https://determined-ai.slack.com/archives/C03LE5Z8R46/p1672776613155459?thread_ts=1672772755.830529&cid=C03LE5Z8R46 I think the idea there is that we probably want to keep it similar to asking for a workspace that doesn't exist (we can handle this situation in a separate pr)
